bee-network/bee-autopeering/src/peer/peer_id.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Peer identities.
+
+use crate::hash;
+
+use crypto::signatures::ed25519::{PublicKey, SecretKey as PrivateKey, PUBLIC_KEY_LENGTH};
+use serde::{
+    de::{SeqAccess, Visitor},
+    ser::SerializeStruct,
+    Deserialize, Serialize,
+};
+
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+};
+
+const DISPLAY_LENGTH: usize = 16;
+
+/// Represents the unique identity of a peer in the network.
+#[derive(Copy, Clone)]
+pub struct PeerId {
+    // An ED25519 public key.
+    public_key: PublicKey,
+    // The SHA256 hash of the ED25519 public key.
+    id_bytes: [u8; hash::SHA256_LEN],
+    // The corresponding `libp2p` PeerId.
+    libp2p_peer_id: libp2p_core::PeerId,
+    // The displayed Base58 representation.
+    base58_peer_id: [char; DISPLAY_LENGTH],
```

Comment:
Does this have any other use case than displaying ? If no, I'd just have a String, otherwise we have to iter/collect every time we display :/

Reply:
`String` is not `Copy` however. Should we give up on `PeerId` being Copy?

Reply:
As I don't want to give up on `PeerId` being `Copy` I leave it as it is for now, but added a NOTE so that we consider changing it in a future patch version.

Reply:
I thought a bit about it, and I think the best way is to not buffer at all (we should instead create the displayed representation JIT), because we copy the `PeerId` much more often than we display it. If logging is disabled we never display, but we'ld still have to copy the display-related bytes, which I don't think is a good idea despite computers being best at copying bytes.

This means that I return this type to the state I had originally :smiley:

Reply:
Done (and thanks to make me aware that I did claim ownership over a few things I didn't have to necessarily, basically I forgot to adapt to some of my refactors that made ownership unnecessary in some intances)